### PR TITLE
Fix a multithreading bug in grpc `ClientCall`

### PR DESCRIPTION
**Pull request by raulchen**
### Problem.
`ClientCallManager` returns a raw pointer of `ClientCall`, and the caller can use `call->GetStatus()` to tell whether the request was successfully sent. However, there's a small chance that the reply is received too fast. And a call may be deleted before `call->GetStatus()`. 

### Solution
This PR fixes this issue by adding a `ClientCallTag` and make `ClientCall` a `shared_ptr`.
Also fixes some typos and some small issues reported by clang-tidy.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

**Comment by raulchen** [comment]
@zhijunfu @jiangzihao2009

**Review by jovany-wang — APPROVED** [review]

File: src/ray/rpc/grpc_server.h [hunk 1]
```diff
@@ -32,8 +33,8 @@ class GrpcServer {
   ///  will be chosen.
   /// \param[in] main_service The main event loop, to which service handler functions
   /// will be posted.
-  GrpcServer(const std::string &name, const uint32_t port)
-      : name_(name), port_(port), is_closed_(true) {}
+  GrpcServer(std::string name, const uint32_t port)
+      : name_(std::move(name)), port_(port), is_closed_(true) {}
```
Comment (pcmoritz) [inline src/ray/rpc/grpc_server.h]: The previous code seemed clearer to me, why did you change it?

Comment (raulchen) [reply src/ray/rpc/grpc_server.h]: Clang-tidy suggests this change. Passing by value + move has some advantages over passing by const reference + copy. 
The biggest advantage is that you can avoid copying data.  E.g.,
```
std::string name = "....";
GrpcServer(std::move(name));  // no copy at all.
```

See https://stackoverflow.com/questions/51705967/advantages-of-pass-by-value-and-stdmove-over-pass-by-reference/51706522 for more details.

Comment (pcmoritz) [reply src/ray/rpc/grpc_server.h]: Ok interesting, thanks for the explanation :)

**Review by pcmoritz — APPROVED** [review]
